### PR TITLE
[SPARK-35133][SQL] Explain codegen works with AQE

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

`EXPLAIN CODEGEN <query>` (and Dataset.explain("codegen")) prints out the generated code for each stage of plan. The current implementation is to match `WholeStageCodegenExec` operator in query plan and prints out generated code (https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala#L111-L118 ). This does not work with AQE as we wrap the whole query plan inside `AdaptiveSparkPlanExec` and do not run whole stage code-gen physical plan rule eagerly (`CollapseCodegenStages`). This introduces unexpected behavior change for EXPLAIN query (and Dataset.explain), as we enable AQE by default now.

The change is to explain code-gen for the current executed plan of AQE.

### Why are the changes needed?

Make `EXPLAIN CODEGEN` work same as before.

### Does this PR introduce _any_ user-facing change?

No (when comparing with latest Spark release 3.1.1).

### How was this patch tested?

Added unit test in `ExplainSuite.scala`.

### CODE REVIEW

**Comment by c21** [comment]
cc @cloud-fan if you have time to take a look, thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala [hunk 1]
```diff
@@ -197,7 +198,14 @@ class QueryExecution(
         queryExecution.toString(maxFields, append)
       case CodegenMode =>
         try {
-          org.apache.spark.sql.execution.debug.writeCodegen(append, queryExecution.executedPlan)
+          queryExecution.executedPlan match {
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala:201]: How about handling the AQE node in the `writeCodegen` side instead? That way, `debugCodegen` can support it, too?

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala]: @maropu - updated, thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala [hunk 2]
```diff
@@ -105,13 +106,20 @@ package object debug {
...
     val codegenSubtrees = new collection.mutable.HashSet[WholeStageCodegenExec]()
 
     def findSubtrees(plan: SparkPlan): Unit = {
       plan foreach {
         case s: WholeStageCodegenExec =>
           codegenSubtrees += s
+        case p: AdaptiveSparkPlanExec =>
+          // Find subtrees from original input plan of AQE.
+          val inputExecutedPlan = QueryExecution.prepareForExecution(
+            QueryExecution.preparations(sparkSession, None), p.inputPlan)
```
Comment (maropu) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:121]: Could we use `SparkSession.getActiveSession` to avoid the unnecessary changes?

**Review by maropu — COMMENTED** [review]
Looks fine otherwise.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala [hunk 3]
```diff
@@ -73,14 +74,14 @@ package object debug {
    * @param plan the query plan for codegen
    * @return single String containing all WholeStageCodegen subtrees and corresponding codegen
    */
-  def codegenString(plan: SparkPlan): String = {
+  def codegenString(plan: SparkPlan, sparkSession: SparkSession): String = {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:77]: shall we simply strip these AQE wrapper nodes at the beginning using `AdaptiveSparkPlanHelper.stripAQEPlan`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala [hunk 4]
```diff
@@ -105,13 +106,20 @@ package object debug {
...
+      plan: SparkPlan,
+      sparkSession: SparkSession): Seq[(String, String, ByteCodeStats)] = {
     val codegenSubtrees = new collection.mutable.HashSet[WholeStageCodegenExec]()
 
     def findSubtrees(plan: SparkPlan): Unit = {
       plan foreach {
         case s: WholeStageCodegenExec =>
           codegenSubtrees += s
+        case p: AdaptiveSparkPlanExec =>
+          // Find subtrees from original input plan of AQE.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:119]: This doesn't match the actual plan. Why can't use `AdaptiveSparkPlanExec.executedPlan`?

Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:119]: This doesn't match the actual plan. Why can't use `AdaptiveSparkPlanExec.executedPlan`?

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:119]: Yes, this doesn't match the actual plan for `df.explain("codegen")` if `df` is executed already. The problem is the final plan `AdaptiveSparkPlanExec.executedPlan` has `ShuffleQueryStageExec` to wrap the whole sub-plan under that shuffle.

Example:


```
spark.range(5).select(col("id").as("key"), col("id").as("value")).groupBy('key).agg(max('value))
```

```
AdaptiveSparkPlan isFinalPlan=true
+- == Final Plan ==
   *(2) HashAggregate(keys=[key#2L], functions=[max(value#3L)], output=[key#2L, max(value)#9L])
   +- CustomShuffleReader coalesced
      +- ShuffleQueryStage 0
         +- Exchange hashpartitioning(key#2L, 5), ENSURE_REQUIREMENTS, [id=#28]
            +- *(1) HashAggregate(keys=[key#2L], functions=[partial_max(value#3L)], output=[key#2L, max#13L])
               +- *(1) Project [id#0L AS key#2L, id#0L AS value#3L]
                  +- *(1) Range (0, 5, step=1, splits=2)
```

The partial aggregate `HashAggregate` is wrapped inside `ShuffleQueryStage`, so cannot be pattern matched to do the explain. One way to workaround is to add pattern matching for `ShuffleQueryStageExec` as well. But anyway we need to re-run the preparation physical plan rules if `AdaptiveSparkPlan.isFinalPlan=false`.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:119]: Yea we can strip `QueryStageExec` as well. Can we use `executedPlan` even if `isFinalPlan=false`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala (hunk 2 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala]: @maropu - good call, it's much cleaner. Updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala (hunk 4 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala]: @cloud-fan - updated.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala [hunk 5]
```diff
@@ -112,6 +113,19 @@ package object debug {
       plan foreach {
         case s: WholeStageCodegenExec =>
           codegenSubtrees += s
+        case p: AdaptiveSparkPlanExec =>
+          // Find subtrees from current executed plan of AQE.
+          val executedPlan = p.executedPlan
+          if (executedPlan.find(_.isInstanceOf[WholeStageCodegenExec]).isEmpty) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:119]: Ah, now I see what you mean. For `isFinalPlan=false`, we may have query stages in the query plan that do not run `queryStageOptimizerRules` yet, and have no `WholeStageCodegenExec`.

I think this is at least consistent with normal EXPLAIN. If the `df` is not run yet, you won't see the leading `*` in plan nodes that indicate whole-stage codegen.

One idea is to trigger AQE during EXPLAIN, so that we can display the final plan. But this won't work for SQL EXPLAIN.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:119]: We need a bit more thinking about how to do EXPLAIN with AQE.

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:119]: @cloud-fan - I feel `SQL EXPLAIN` is usable, at least for me and several developers in the team. It's kind of annoying that we need to disable AQE every time for this.

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:119]: > One idea is to trigger AQE during EXPLAIN, so that we can display the final plan.

I think users and developers are expecting `EXPLAIN` to be a purely metadata operator, so it might be surprising or inefficient to execute actual query during `EXPLAIN`.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:119]: After some plan changes (such as join elimination) in AQE, I don't think SQL EXPLAIN is useful as you can never see the final plan. How do you expect SQL EXPLAIN to work for AQE?

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:119]: Because AQE is not changing every query. As an example I gave in unit test case in `ExplainSuite.scala`, for aggregation query, AQE should not affect the behavior for the generated code for aggregation.

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:119]: I am fine with changing the code to be `findSubtrees(aqePlan.executedPlan)`, so the explain code-gen behavior is same as explain, and at least it fixes the case for data frame being executed already (is final plan).

I can just add the change for `SQL EXPLAIN` in our internal fork and discuss in the future. WDYT @cloud-fan ?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala:119]: SGTM. We can also look at how other systems do EXPLAIN with query plan that can change at runtime.

Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/debug/package.scala]: @cloud-fan - updated, I agree, will check as well, thanks.

**Review by cloud-fan — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 6]
```diff
@@ -565,6 +565,33 @@ class ExplainSuiteAE extends ExplainSuiteHelper with EnableAdaptiveExecutionSuit
          |""".stripMargin
     )
   }
+
+  test("SPARK-35133: explain codegen should work with AQE") {
+    withSQLConf(SQLConf.WHOLESTAGE_CODEGEN_ENABLED.key -> "true") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:570]: Hi, @c21 . 
Although SPARK-33679 enabled AQE by default for Apache Spark 3.2, we don't know what happens during QA and Apache Spark 3.2.0 RC period. In addition, we had better an explicit configuration in the test case. Could you add `spark.sql.adaptive.enabled=true` explicitly?

Comment (c21) [reply sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:570]: @dongjoon-hyun - sorry if it's not clear, this test is under `ExplainSuiteAE` which extends `EnableAdaptiveExecutionSuite` to enforce AQE enabled already.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala [hunk 7]
```diff
@@ -24,14 +24,13 @@ import org.apache.spark.sql.catalyst.InternalRow
...
 import org.apache.spark.sql.catalyst.expressions.codegen.CodegenContext
 import org.apache.spark.sql.execution.{CodegenSupport, LeafExecNode, WholeStageCodegenExec}
-import org.apache.spark.sql.execution.adaptive.DisableAdaptiveExecutionSuite
+import org.apache.spark.sql.execution.adaptive.{DisableAdaptiveExecutionSuite, EnableAdaptiveExecutionSuite}
 import org.apache.spark.sql.functions._
 import org.apache.spark.sql.test.SharedSparkSession
 import org.apache.spark.sql.test.SQLTestData.TestData
 import org.apache.spark.sql.types.StructType
 
-// Disable AQE because the WholeStageCodegenExec is added when running QueryStageExec
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala:33]: Instead of removal, shall we recover this for `DebuggingSuite` at line 100?

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (with minor comments). Thank you, @c21 and all.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala (hunk 7 above)
Comment (c21) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala:33]: @dongjoon-hyun - sorry can you elaborate more by recovering? Here I kept the AQE-runnable unit tests under `DebuggingSuiteBase`, and have a non-AQE `DebuggingSuite` and AQE `DebuggingSuiteAE`.

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala (hunk 6 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:570]: Oh, got it. Then, it's perfect.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala (hunk 7 above)
Comment (c21) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala:33]: oh do you mean move the comment to line 100?

Comment (c21) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala:33]: @dongjoon-hyun - updated to move the comment to line 100, let me know if my understanding is not correct, thanks.

**Review by maropu — APPROVED** [review]

**Review by imback82 — APPROVED** [review]
LGTM

File: sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala (hunk 7 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/debug/DebuggingSuite.scala:33]: Thanks!

**Comment by dongjoon-hyun** [comment]
It seems that there is some delay at GitHub Action. I checked that it's already passed.
Screen Shot 2021-05-06 at 8 42 37 PM

Thank you, @c21 and all. Merged to master.

**Comment by c21** [comment]
Thank you all for review!
